[Server] Add ability to set avatar

katevi opened the pull request:
User should be able to have an avatar.
The avatar saving alhorythm:
1. User clicks "Set profile picture" button from front-end;
2. Loads the picture
3. Clicks "save changes";
4. The picture is going to the persistent storage.
If user hasn't set an avatar, a default one (for example: the ? picture) is assigned.

Note that it'd require changing of user entity. Information about the user should also contain its avatar. #20

Also we shouldn't have a separate method in order to receive a response. registration/user and login/user should return a response themselves. #22

AndreyKoltsov1997 commented on chat-server/src/main/java/com/vinnik/chat/back/perstistence/RegistrationController.java:
```diff
@@ -28,23 +31,53 @@ public User getUserByUserNickname(@PathVariable("nickname") String nickname) {
+
             userService.saveOrUpdateUser(user);
-            return new ResponseEntity("User added successfully", HttpStatus.OK);
+            return userData;
```
I'd return JSON containing user information (entity).

AndreyKoltsov1997 commented on chat-server/src/main/java/com/vinnik/chat/back/perstistence/RegistrationController.java:
```diff
@@ -28,23 +31,53 @@ public User getUserByUserNickname(@PathVariable("nickname") String nickname) {
         } catch (NicknameAlreadyExistsException e) {
             e.printStackTrace();
-            return new ResponseEntity<>("Nickname already exists", HttpStatus.CONFLICT);
+            return null;
```
I don't think returning null value in case of an error is sufficient. Ultimately, the user should be aware of the reason of the error has occurrence.

AndreyKoltsov1997 commented on chat-server/src/main/java/com/vinnik/chat/back/perstistence/RegistrationController.java:
```diff
@@ -28,23 +31,53 @@ public User getUserByUserNickname(@PathVariable("nickname") String nickname) {
+     * Set avatar to user.
+     * @param nickname - name of user, whose avatar to set.
+     * @param file - user's avatar.
+     * @return - if avatar set successfully returns ok status, if file is empty returns "not found" status
```
Not sure if statement `if file is empty returns "not found" status` actually represent the behaviour. I assume it returns status "not found" in case any error has occurred.

katevi commented:
> 1. I'd consider to change some of the endpoints:
> 
> * GET `/registration/users/` -> `/register`;
> * GET `/registration/users/{nickname}`-> `/user?nickname=...`. I don't really understand the reason you use a separate endpoints instead of HTTP query variables;
> * DELETE `/registration/users/{nickname}`-> `/user?nickname=...`;
> * POST `/login/user/` -> '/login'. I assume only users could log in :)
> 
> * remove "app" prefix where possible. Example: `/app/chat.newUser` -> `/chat.newUser`;
> 
> 1. I think it'd be convenient to add ability to register user with the avatar right away while still keeping a separate setAvatar(...) method.

I made separate setAvatar(...) method, since I thought that if user entered not valid registration data, picture will be sended to server uselessly. Also I do not know how to send files and user entity in one query (I googled already).

AndreyKoltsov1997 commented:
> > 1. I'd consider to change some of the endpoints:
> > 
> > 
> > * GET `/registration/users/` -> `/register`;
> > * GET `/registration/users/{nickname}`-> `/user?nickname=...`. I don't really understand the reason you use a separate endpoints instead of HTTP query variables;
> > * DELETE `/registration/users/{nickname}`-> `/user?nickname=...`;
> > * POST `/login/user/` -> '/login'. I assume only users could log in :)
> > * remove "app" prefix where possible. Example: `/app/chat.newUser` -> `/chat.newUser`;
> > 
> > 
> > 1. I think it'd be convenient to add ability to register user with the avatar right away while still keeping a separate setAvatar(...) method.
> 
> I made separate setAvatar(...) method, since I thought that if user entered not valid registration data, picture will be sended to server uselessly. Also I do not know how to send files and user entity in one query (I googled already).

I assume it's possible via encoding the file into Base64.
But, I've understood your point. In our case, it's not critical and we could keep the registration arguments the same.

AndreyKoltsov1997 commented on chat-server/README.md:
```diff
@@ -13,14 +13,11 @@ Pet-project of participant Dell mentorship program. Made with Java.
+| `/topic/publishedMessages` | **GET** | - | - | Server posts messages, received from clients |
+| `/registration/users/` | **GET** | - | - | Get all registered users |
+| `/registration/users/{nickname}`| **GET** | - | nickname - nickname of registered user | Get user by specialized nickname |
+| `/registration/users/{nickname}` | **DELETE** | - | nickname - nickname of registered user | Delete user by specialized nickname |
```
I assume `nickname` is now a `/registration/users?...` parameter.

AndreyKoltsov1997 commented on chat-server/README.md:
```diff
@@ -13,14 +13,11 @@ Pet-project of participant Dell mentorship program. Made with Java.
-| `login/user/` | **POST** | `application/json` | `nickname` - nickname of user, who want to log in, `password` - user's password | Tries to log in user with specialized metadata |
-| `login/user/response` | **POST** | `message/http` | - | Sends http status code about result of login |
-
+| `/app/sentMessages` | **POST** | `application/json` | `message` - text of user's message, `sender` - user's nickname |  Server receives chat message from registered user |
```
Does `/app/` still exist?

katevi commented:
1) When I removed "/registration/" endpoint from the RegistrationController to achieve endpoint like "/user/?nickname=...", I received mistake: "method get not supported".
I googled and found [answer](https://stackoverflow.com/questions/31570021/failed-to-create-websocket-connection-when-spring-security-is-on/56715980), which says, that it is impossible to have Controller without requestMapping for it.
So, as "/registration" endpoint is not actual now, I joined RegistrationController and LoginController into UserController and added method "/api", like in [this](https://www.baeldung.com/spring-request-method-not-supported-405) tutorial. "/api" prefixes underlines that there is REST API in my app.
So, it seems impossible for me to have only "/user..", I added prefix "api/" for all methods.
2) I have googled and found [tips](https://florimond.dev/blog/articles/2018/08/restful-api-design-13-best-practices-to-make-your-users-happy/) about writing good API. Author says it is bad practice to write verb in API, so, I changed "/register" to "/registered" in get method. It is also more correctly shows, that we want to receive list of registered users.

AndreyKoltsov1997 commented:
>    So, as "/registration" endpoint is not actual now, I joined RegistrationController and LoginController into UserController and added method "/api", like in [this](https://www.baeldung.com/spring-request-method-not-supported-405) tutorial. "/api" prefixes underlines that there is REST API in my app.

It totally makes sense to add `/api/` prefix when you have logic for both front-end and back-end within the same server, so It'd be clearer to classify methods that handles REST API and methods that does front-end file serving.
Our case is different - we have a separate server for front-end serving. If it'd had a REST API, then it'd be better to add `/api/` prefix there.
